helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java

```diff
@@ -217,6 +218,23 @@ public void requireFullRefresh() {
     }
   }
 
+  /**
+   * Update cache and return true if any local data is changed compared to before refresh.
+   * @param dataAccessor the data accessor used to fetch data.
+   * @return true if there is a change to local cache.
+   */
+  public boolean updateCache(HelixDataAccessor dataAccessor) {
+    Map<String, LiveInstance> liveInstance = new HashMap<>(getLiveInstances());
+    Map<String, ExternalView> externalView = new HashMap<>(getExternalViews());
+    Map<String, InstanceConfig> instanceConfig = new HashMap<>(getInstanceConfigMap());
+    requireFullRefresh();
+    refresh(dataAccessor);
+
+    return !(liveInstance.equals(getLiveInstances())
```

Comment:
i don't get this. you assigned liveInstace = getLiveInstances() on line: 227. and on line: 233 - you are again checking - which is just few cpu cycles after the first call

Reply:
The first one is to get a snapshot, and compare with the one after a refresh.

Reply:
but snapshot is taken just few cpu cycles before on line-227

Reply:
refresh() happen and update the cache data in the middle of two `getLiveInstances()`, the data may be changed.